WDPD-34 Credit Card Payment Method

ps-cgrach opened the pull request:
### This PR

Adds Credit card payment method

### Jira Tickets

https://jira.parkside.at/browse/WDPD-34
https://jira.parkside.at/browse/WDPD-112

ps-holler commented on Extend/Controller/OrderController.php:
```diff
@@ -128,8 +162,35 @@ public function execute()
+     */
+    private function _handleFormResponse($oSession, $oPayment, $oOrder)
+    {
+        if (!empty($oResponse = $oSession->getVariable(self::FORM_POST_VARIABLE))) {
```
I know OXID does this as well, but I don't think we should do variable assignments in `if` statements. It's generally considered an anti-pattern.

ps-cgrach replied:
👍

ps-holler commented on Extend/Controller/OrderController.php:
```diff
@@ -246,14 +316,43 @@ private function _handleResponse($oResponse, $oLogger)
+
+            $sSid = $oSession->sid(true);
+            if ($sSid != '') {
+                $sSid = '&' . $sSid;
```
You can also use `http_build_query` for this.

ps-cgrach replied:
no, `$sSid` is already a query

ps-holler commented on Extend/Controller/OrderController.php:
```diff
@@ -273,10 +369,126 @@ private function _isPaymentFinished($oSessionToken, $oPaymentRedirect)
+        $oSession = $this->getSession();
+        $sSid = $oSession->sid(true);
+        if ($sSid != '') {
+            $sSid = '&' . $sSid;
```
You can also use `http_build_query` for this.

ps-cgrach replied:
:+1:

ps-holler commented on Extend/Controller/OrderController.php:
```diff
@@ -273,10 +369,126 @@ private function _isPaymentFinished($oSessionToken, $oPaymentRedirect)
+     *
+     * @return string
+     */
+    private function _getPaymentAction(string $sAction): string
```
Is this specific to credit card? If not, we might want to put it somewhere else.

ps-cgrach replied:
Actually, it is specific to the Javascript form provided by the payment SDK. So every other Payment method with that behavior will use it. I think we will not need it anywhere else.

ps-holler commented on Model/CreditCardPaymentMethod.php:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * @inheritdoc
+     *
+     * @return Transaction
```
```suggestion
     * @return CreditCardTransaction
```

If you use that you can omit `use Wirecard\PaymentSdk\Transaction\Transaction` above.

ps-cgrach replied:
👍

ps-holler commented on Model/CreditCardPaymentMethod.php:
```diff
@@ -0,0 +1,189 @@
+
+        $oConfig->add($oCreditCardConfig);
+
+        $this->_oLogger->debug("CreditCardPaymentMethod#getConfig():" . print_r($oConfig, true));
```
I think you don't need to put the method or class name here – the logger should have access to the stack anyway.

ps-cgrach replied:
removed, was just debugging output

ps-holler commented on Model/CreditCardPaymentMethod.php:
```diff
@@ -0,0 +1,189 @@
+        $oThreeDCurrency = $oShopConfig->getCurrencyObject($oPayment->oxpayments__wdoxidee_limits_currency->value);
+        $oShopCurrency = $oShopConfig->getActShopCurrencyObject();
+
+        if ($oPayment->oxpayments__wdoxidee_non_three_d_max_limit->value !== '') {
```
```suggestion
        if ($oPayment->oxpayments__wdoxidee_non_three_d_max_limit->value) {
```
I suppose the empty string check is used so that `'0'` is included? If so, does it really make sense to set a limit of 0? This also applies for the min limit below.

ps-cgrach replied:
yes, Applying a 0 limit will behave differently than not setting a limit at all.

ps-holler commented on Model/PaymentGateway.php:
```diff
@@ -7,7 +7,7 @@
  */
 
-namespace Wirecard\Oxid\Extend\Model;
+namespace Wirecard\Oxid\Model;
```
Why? This still extends a model (`OxidEsales\EshopCommunity\Core\Model\BaseModel`).

ps-cgrach replied:
Ah, yes, the extension was added later again.

ps-holler commented on Model/PaymentGateway.php:
```diff
@@ -71,44 +67,37 @@ public function __construct()
 
         $oRedirect = new Redirect(
-            $sShopUrl . 'index.php?lang=' . $sBaseLanguage . '&cl=order&wdpayment=' . $sRandom . $sSid,
+            $sShopUrl . 'index.php?lang=' . $sBaseLanguage . '&cl=order' . $sModuleToken . $sSid,
```
You should use `http_build_query` for this.

ps-holler commented on Model/PaymentGateway.php:
```diff
@@ -119,72 +108,117 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
-        $oRedirect = self::getRedirectUrls($oSession, $sShopUrl);
-        $oPaymentMethod = Payment_Method_Factory::create($sPaymentMethod);
-        $oTransactionService = new TransactionService($oPaymentMethod->getConfig(), $this->_oLogger);
+        /*
```
```
 *
*
*
```
:/

ps-cgrach replied:
😅

ps-holler commented on Model/PaymentGateway.php:
```diff
@@ -119,72 +108,117 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+
+        $sSid = $oSession->sid(true);
+        if ($sSid != '') {
+            $sSid = '&' . $sSid;
```
You should use… `http_build_query`. 🎉

ps-holler commented on Model/PaymentGateway.php:
```diff
@@ -119,72 +108,117 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+
+        $sSid = $oSession->sid(true);
+        if ($sSid != '') {
+            $sSid = '&' . $sSid;
```
Alright. `http_build_query`

ps-holler commented on Model/PaymentGateway.php:
```diff
@@ -119,72 +108,117 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+            $sBaseLanguage = Registry::getLang()->getBaseLanguage();
+            $oResponse = $oTransactionService->processJsResponse(
+                $_POST,
+                $sShopUrl . 'index.php?lang=' . $sBaseLanguage . '&cl=order&redirectFromForm=1' . $sModuleToken . $sSid
```
`http`
`_`
`build`
`_`
`query`

ps-holler commented on Tests/Unit/Core/Payment_Method_Factory_Test.php:
```diff
@@ -0,0 +1,26 @@
+    public function testCreatePaypal()
+    {
+        $oPaymentMethod = Payment_Method_Factory::create("wdpaypal");
+        $this->assertTrue($oPaymentMethod instanceof \Wirecard\Oxid\Model\Paypal_Payment_Method);
```
Just a minor note: There is also `assertInstanceOf()`.

ps-cgrach replied:
nice!

ps-holler commented on Tests/Unit/Model/Paypal_Payment_Method_Test.php:
```diff
@@ -21,12 +23,14 @@ class Paypal_Payment_Method_Test extends OxidEsales\TestingLibrary\UnitTestCase
     {
         parent::setUp();
-        $this->oPaymentMethod = new Paypal_Payment_Method;
+        $this->oPaymentMethod = new Paypal_Payment_Method();
```
👍

ps-holler commented on metadata.php:
```diff
@@ -57,14 +57,12 @@
             => \Wirecard\Oxid\Extend\Model\Basket::class,
         \OxidEsales\Eshop\Application\Controller\ThankYouController::class
-            => \Wirecard\Oxid\Extend\Controller\ThankYouController::class
+            => \Wirecard\Oxid\Extend\Controller\ThankYouController::class,
```
Finally, I'm not alone. 👍

ps-holler commented on out/blocks/wirecard_credit_card_fields.tpl:
```diff
@@ -0,0 +1,30 @@
+[{$smarty.block.parent}]
+[{assign var="payment" value=$oView->getPayment()}]
+[{if $payment->oxpayments__oxid->value eq "wdcreditcard"}]
```
`eq` is interesting, but so far we have always simply used `==` (or, even better, `===`).

ps-cgrach replied:
it was not working before with `===`, now it does. maybe it was another syntax error then.

ps-holler commented on out/js/credit_card_form.js:
```diff
@@ -0,0 +1,85 @@
+    form.submit();
+  }
+
+  return obj = {
```
```suggestion
  return {
```

`obj` is not used anywhere.

ps-cgrach replied:
👍

ps-holler commented on out/js/credit_card_form.js:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  function setParentTransactionId(response) {
+    console.log("setParentTransactionId", response);
```
Debugging leftover?

ps-cgrach replied:
:+1:

ps-holler commented on out/js/credit_card_form.js:
```diff
@@ -0,0 +1,85 @@
+      orderButton.prop("disabled", true);
+      orderButton.on("click", function (event) {
+        event.preventDefault();
+        submitPaymentForm(event)
```
```suggestion
        submitPaymentForm(event);
```

ps-cgrach replied:
👍

ps-holler commented on out/js/credit_card_form.js:
```diff
@@ -0,0 +1,85 @@
+
+  function logError(where, error) {
+    if (typeof error == "string") {
+      $("#wirecard-cc-error").html(error)
```
```suggestion
      $("#wirecard-cc-error").html(error);
```

ps-cgrach replied:
:+1:

ps-holler commented on out/js/credit_card_form.js:
```diff
@@ -0,0 +1,85 @@
+ * - License can be found under:
+ * https://github.com/wirecard/oxid-ee/blob/master/LICENSE
+ */
+var CreditCardForm = (function ($) {
```
We might want to prefix global vars to avoid conflicts with core and other modules.

ps-cgrach replied:
👍

ps-holler commented on out/blocks/wd_admin_payment_main_form.tpl:
```diff
@@ -1,7 +1,7 @@
   [{assign var="paymentMethod" value=$edit->getPaymentMethod()}]
   [{assign var="configFields" value=$paymentMethod->getConfigFields()}]
-[{/if}]
+  [{/if}]
```
Why the indentation changes in this file?

ps-cgrach replied:
maybe because of the missing closing tags for the logo and description `<tr>`'s

ps-holler commented on out/blocks/wd_admin_payment_main_form.tpl:
```diff
@@ -57,71 +57,86 @@
-[{/if}]
+  [{if $edit->oxpayments__oxid->value === 'wdcreditcard'}]
+  <tr>
+    <td></td>
```
You can use `colspan="2"` here instead of creating two `<td>` elements.

ps-holler commented on out/blocks/wd_admin_payment_main_form.tpl:
```diff
@@ -57,71 +57,86 @@
+  [{/if}]
   [{/foreach}]
-[{/if}]
+  [{if $edit->oxpayments__oxid->value === 'wdcreditcard'}]
```
I think it would be cleaner if this was in `Payment_Method::getConfigFields`. You could create a new field type for this, e.g. "info".

ps-cgrach replied:
:+1:

ps-cgrach commented on Controller/FormInteractionController.php:
```diff
@@ -0,0 +1,70 @@
+class FormInteractionController extends FrontendController
+{
+    /**
+     * @inheridoc
```
👍

tomazpu commented on Extend/Controller/OrderController.php:
```diff
@@ -286,4 +377,118 @@ private function _isFinalizeOrderSuccessful($iSuccess)
+        ));
+        $oTransaction->setConfig($this->_getCreditCardPaymentMethodConfig()->get(CreditCardTransaction::NAME));
+
+        $oSession = $this->getSession();
```
put this into a separate method to much logic happens in this method

ps-cgrach replied:
👍

tomazpu commented on Extend/Controller/OrderController.php:
```diff
@@ -286,4 +377,118 @@ private function _isFinalizeOrderSuccessful($iSuccess)
+        $oPayment = oxNew(Payment::class);
+        $oPayment->load($oBasket->getPaymentId());
+
+        return "ModuleCreditCardForm.init(" .
```
Comment that this gets loaded in the template and outputs the data to be used in javascript

tomazpu commented on Extend/Controller/OrderController.php:
```diff
@@ -286,4 +377,118 @@ private function _isFinalizeOrderSuccessful($iSuccess)
+    private function _getTransactionService(): TransactionService
+    {
+        if (is_null($this->_oTransactionService)) {
+            $this->_oTransactionService = new TransactionService($this->_getCreditCardPaymentMethodConfig());
```
Logger is missing

tomazpu commented on Extend/Controller/OrderController.php:
```diff
@@ -286,4 +377,118 @@ private function _isFinalizeOrderSuccessful($iSuccess)
+     */
+    private function _getPaymentAction(string $sAction): string
+    {
+        if ($sAction == 'pay') {
```
please use paymentSDK constants for the purchase and authorization. Also i think there is one for pay

ps-cgrach replied:
👍

tomazpu commented on Extend/Model/Order.php:
```diff
@@ -246,4 +248,19 @@ public function getTranslatedState()
+     * Create a temporary Order that should not be saved but used for creating a
+     * {@link Wirecard\PaymentSdk\Transaction\Transaction}
+     *
+     * @param Basket $oBasket
```
Basket is undefined and User too.

ps-cgrach replied:
they are, see line 14 && 15

tomazpu commented on Extend/Model/Order.php:
```diff
@@ -246,4 +248,19 @@ public function getTranslatedState()
+    {
+        $this->_setUser($oUser);
+        $this->_loadFromBasket($oBasket);
+        $this->oxorder__oxid = new Field(Registry::getSession()->getVariable('sess_challenge'));
```
The session challenge is the oxid order id?

ps-cgrach replied:
yes, see `OxidEsales\Eshop\Application\ModelOrder#finalizeOrder`, that is the class we extend, and there they use the `sess_challenge` for the order id.

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -71,44 +68,37 @@ public function __construct()
      * @SuppressWarnings(PHPMD.Coverage)
      */
-    public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+    public static function getRedirectUrls($oSession, $sShopUrl, $sSid): Redirect
```
Why not use the same logic with addRedirectUrls, always keep in mind that the logic and coding standard is consistent and easy to understand.

ps-cgrach replied:
👍

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+    public function createTransaction($oBasket, $oOrder)
     {
         $sShopUrl = $this->getConfig()->getCurrentShopUrl();
+
```
I think there i no need for the newline

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+        $oPayment = oxNew(Payment::class);
+        $oPayment->load($oBasket->getPaymentId());
+
+        $sSid = $oSession->sid(true);
```
if you have a helper you can use the same method as above in another class for this

ps-cgrach replied:
👍

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
-        $oBasket = $oSession->getBasket();
+        $oTransaction->setAmount(new Amount($oBasket->getPrice()->getBruttoPrice(), $oCurrency->name));
 
         $sOrderDetails = $oOrder->oxorder__oxremark->value;
```
put the order details into an addOrderDetails and if its not empty you add it. You can also create a addMandatoryTransactionData() where you add notification url, redirects, amount, order detail.

ps-cgrach replied:
👍

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+        $oShopConfig = Registry::getConfig();
+        $sShopUrl = $oShopConfig->getCurrentShopUrl();
+
+        $sSid = $oSession->sid(true);
```
again the same logic

ps-cgrach replied:
👍

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+    public function executeTransaction($oTransaction, $oOrder)
+    {
+        $oSession = $this->getSession();
+        $oBasket = $oSession->getBasket();
```
hm.. you using session again only for the sSid, extract it. And the basket only to get the paymentId, can u send it in the method. You need in both execute and create transaction the basket so just create the basket in handleTransaction and send it.

ps-cgrach replied:
👍

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+            $sSid = '&' . $sSid;
+        }
+
+        if (!empty($oShopConfig->getRequestParameter('jsresponse'))) {
```
make to methods one to process a "jsResponse" and one for processing "normal responses" then you can create just a isJSResponse and check for it then use one of the new methods to process.

tomazpu replied:
please update according to the first review

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+            $sModuleToken = self::getModuleToken($oSession);
+            $sBaseLanguage = Registry::getLang()->getBaseLanguage();
+            $oResponse = $oTransactionService->processJsResponse(
+                $_POST,
```
There is a method to get $_POST and the global $_POSTs are usually not be used

ps-cgrach replied:
Are you sure there is somenthing. OXID does only has the `getRequestParameter` but you need the name of the parameter. the `processJsResponse` expects an array and creates a response out of the provided values. I think I should not make any assumptions on the values it expects and create my own array.

tomazpu replied:
ok

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+            );
+            if ($oResponse instanceof SuccessResponse) {
+                $oBackendService = new BackendService($oTransactionConfig, $this->_oLogger);
+                NotifyHandler::onSuccessResponse($oResponse, $oBackendService, $oOrder);
```
again misleading that a notifayHandler processes responses

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -119,63 +109,114 @@ public function getRedirectUrls(Session $oSession, string $sShopUrl): Redirect
+    {
+        $oSession = $this->getSession();
+        $oBasket = $oSession->getBasket();
+        /*
```
one star missing in the comment

tomazpu commented on Model/CreditCardPaymentMethod.php:
```diff
@@ -0,0 +1,193 @@
+use Wirecard\PaymentSdk\Config\CreditCardConfig;
+use Wirecard\PaymentSdk\Entity\Amount;
+use Wirecard\PaymentSdk\Transaction\CreditCardTransaction;
+use Wirecard\PaymentSdk\Transaction\Transaction;
```
Transaction is not used anywhere in the class

tomazpu commented on Model/Payment_Method.php:
```diff
@@ -41,7 +43,6 @@ abstract class Payment_Method
      *
-     * @SuppressWarnings(PHPMD.Coverage)
      */
     public function __construct()
```
the constructor has a comment thats a paypal constructor

ps-cgrach commented on Controller/NotifyHandler.php:
```diff
@@ -17,7 +17,7 @@
 use \Wirecard\PaymentSdk\Exception\MalformedResponseException;
```
moved response handling to new `ResponseHandler`

ps-cgrach commented on Extend/Controller/OrderController.php:
```diff
@@ -286,4 +377,118 @@ private function _isFinalizeOrderSuccessful($iSuccess)
+        $oPaymentGateway = oxNew(PaymentGateway::class);
+
+        /**
+         * @var $oBasket Basket
```
there is, `use OxidEsales\Eshop\Application\Model\Basket;` Line 14

ps-cgrach commented on Extend/Controller/OrderController.php:
```diff
@@ -286,4 +377,118 @@ private function _isFinalizeOrderSuccessful($iSuccess)
+     * @return string
+     * @throws \Exception
+     */
+    public function getInitCreditCardFormJavaScript(): string
```
It really returns, as a string, JavaScript that is injected in the template.

```
return "ModuleCreditCardForm.init(" .
            $oTransactionService->getCreditCardUiWithData(
                $oTransaction,
                $this->_getPaymentAction($oPayment->oxpayments__wdoxidee_transactionaction->value),
                Registry::getLang()->getLanguageAbbr()
            ) . ")";
```

ps-cgrach commented on Core/Payment_Method_Factory.php:
```diff
@@ -9,7 +9,7 @@
-use Wirecard\Oxid\Model\Credit_Card_Payment_Method;
+use Wirecard\Oxid\Model\CreditCardPaymentMethod;
 use Wirecard\Oxid\Model\Payment_Method;
 use Wirecard\Oxid\Model\Paypal_Payment_Method;
```
fixed!

ps-ohanel commented on out/blocks/wirecard_credit_card_fields.tpl:
```diff
@@ -0,0 +1,38 @@
+[{*
```
Could you move the file into views/blocks

ps-cgrach replied:
removed, also the remaining items in `/out/blocks` are now in the `/views` dir

ps-ohanel commented on out/blocks/wd_admin_payment_main_form.tpl:
```diff
@@ -1,7 +1,7 @@
 [{if $edit && $edit->isCustomPaymentMethod()}]
```
Could you move the file into views/blocks?

ps-ohanel commented on Controller/FormInteractionController.php:
```diff
@@ -0,0 +1,70 @@
+        $this->_oResponse = Registry::getSession()->getVariable('wdFormInteractionResponse');
+        if (empty($this->_oResponse)) {
+            //redirect home
+            Registry::getUtils()->redirect(Registry::getConfig()->getShopUrl() . 'index.php');
```
There is getShopHomeUrl() method in config. Could be used here?

ps-cgrach replied:
nice! thanks for the hint!

tomazpu commented on Extend/Model/PaymentGateway.php:
```diff
@@ -0,0 +1,330 @@
+    /**
+     * Returns a redirect object
+     *
+     * @param Transaction $oTransaction the wirecard transaction
```
Please remove wirecard

tomazpu commented on Model/CreditCardPaymentMethod.php:
```diff
@@ -0,0 +1,192 @@
+use OxidEsales\Eshop\Core\Registry;
+use Wirecard\Oxid\Core\Helper;
+
+use Wirecard\PaymentSdk\Config\Config as Wirecard_Config;
```
Please rename according to the standard, i suggest PaymentSdkConfig

ps-cgrach replied:
👍

tomazpu commented:
I found that the TransactionList.php requires Wirecard\Oxid\Model\Transaction;  which is not used. Please remove it. Also add an Order folder in Controller/Admin/* to group all Order tabs. What i noticed is that the phpdoc is missing a @since tag. In Core i suggest also creating a helper folder. I still see that you use sometimes type hinting in the phpdoc then other times you do it in the method deceleration. Please be consistent. Other than that looks good. Great job!

ps-cgrach commented on Core/ResponseHandler.php:
```diff
@@ -0,0 +1,140 @@
+        $oPayment = oxNew(Payment::class);
+        $oPayment->load($sPaymentMethod);
+
+        $oUtilsDate = Registry::getUtilsDate();
```
👍

ps-cgrach commented on Core/ResponseHandler.php:
```diff
@@ -0,0 +1,140 @@
+        $oTransaction->wdoxidee_ordertransactions__currency = new Field($aData['currency']);
+
+        $oTransaction->wdoxidee_ordertransactions__responsexml = new Field(base64_encode($oResponse->getRawData()));
+        $oTransaction->wdoxidee_ordertransactions__date = new Field($sConvertedTimestamp);
```
👍

ps-cgrach commented on Core/ResponseHandler.php:
```diff
@@ -0,0 +1,140 @@
+            $oResponse->getData()['statuses.0.provider-transaction-id']
+        );
+        $oOrder->oxorder__wdoxidee_transactionid = new Field($oResponse->getTransactionId());
+        $oOrder->oxorder__oxpaid = new Field($sConvertedTimestamp);
```
👍

ps-cgrach commented on Core/ResponseHandler.php:
```diff
@@ -0,0 +1,140 @@
+
+        $oOrder->oxorder__wdoxidee_orderstate
+            = new Field($oBackendService->getOrderState($oResponse->getTransactionType()));
+        $oOrder->oxorder__wdoxidee_final = new Field(1);
```
no, that is for internal behavior. it is not that the transaction was final.

tomazpu replied:
👍

ps-cgrach commented:
- TransactionList.php uses Wirecard\Oxid\Model\Transaction on line 26, 47 and 48.
- moved the Order* classes.
- for type hinting, we will create a code sniffer that checks that we only use the Phpdoc. @see WDPD-111.
- for the `@since` tag we added another ticket (WDPD-113). It is a little late to mention that after having like 40 classes without that tag.

tomazpu commented:
1. I guess i did't see it.
2. Great.
3. Ok.
4. As it is a string replace i don't think i causes that much work.
